In [4]:
import pandas as pd
import numpy as np
import matplotlib as plt
import warnings
warnings.filterwarnings("ignore")

In [5]:
df_features = pd.read_csv("SpotifyFeatures.csv")
df_tracks = pd.read_csv("tracks.csv")

In [6]:
df_tracks.head()

id                                 name  popularity  \
0  35iwgR4jXetI318WEWsa1Q                                Carve           6   
1  021ht4sdgPcrDgSk7JTbKY  Capítulo 2.16 - Banquero Anarquista           0   
2  07A5yehtSnoedViJAZkNnc   Vivo para Quererte - Remasterizado           0   
3  08FmqUhxtyLTn6pAh6bk45        El Prisionero - Remasterizado           0   
4  08y9GfoqCWfOGsKdwojr5e                  Lady of the Evening           0   

   duration_ms  explicit              artists                  id_artists  \
0       126903         0              ['Uli']  ['45tIt06XoI0Iio4LBEVpls']   
1        98200         0  ['Fernando Pessoa']  ['14jtPCOoNZwquk5wd9DxrY']   
2       181640         0  ['Ignacio Corsini']  ['5LiOoJbxVSAMkBS2fUm3X2']   
3       176907         0  ['Ignacio Corsini']  ['5LiOoJbxVSAMkBS2fUm3X2']   
4       163080         0      ['Dick Haymes']  ['3BiJGZsyX9sJchTqcSA7Su']   

  release_date  danceability  energy  key  loudness  mode  speechiness  \
0   1922-02-22         0.645  0.4450    0   -13.338     1       0.4510   
1   1922-06-01         0.695  0.2630    0   -22.136     1       0.9570   
2   1922-03-21         0.434  0.1770    1   -21.180     1       0.0512   
3   1922-03-21         0.321  0.0946    7   -27.961     1       0.0504   
4         1922         0.402  0.1580    3   -16.900     0       0.0390   

   acousticness  instrumentalness  liveness  valence    tempo  time_signature  
0         0.674            0.7440     0.151    0.127  104.851               3  
1         0.797            0.0000     0.148    0.655  102.009               1  
2         0.994            0.0218     0.212    0.457  130.418               5  
3         0.995            0.9180     0.104    0.397  169.980               3  
4         0.989            0.1300     0.311    0.196  103.220               4

In [7]:
df_tracks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 586672 entries, 0 to 586671
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                586672 non-null  object 
 1   name              586601 non-null  object 
 2   popularity        586672 non-null  int64  
 3   duration_ms       586672 non-null  int64  
 4   explicit          586672 non-null  int64  
 5   artists           586672 non-null  object 
 6   id_artists        586672 non-null  object 
 7   release_date      586672 non-null  object 
 8   danceability      586672 non-null  float64
 9   energy            586672 non-null  float64
 10  key               586672 non-null  int64  
 11  loudness          586672 non-null  float64
 12  mode              586672 non-null  int64  
 13  speechiness       586672 non-null  float64
 14  acousticness      586672 non-null  float64
 15  instrumentalness  586672 non-null  float64
 16  liveness          58

In [8]:
df_tracks.isnull().sum()

id                   0
name                71
popularity           0
duration_ms          0
explicit             0
artists              0
id_artists           0
release_date         0
danceability         0
energy               0
key                  0
loudness             0
mode                 0
speechiness          0
acousticness         0
instrumentalness     0
liveness             0
valence              0
tempo                0
time_signature       0
dtype: int64

In [9]:
df_tracks.duplicated().sum()

0

In [10]:
df_tracks.value_counts(["artists"])

artists                                                            
['Die drei ???']                                                       3856
['TKKG Retro-Archiv']                                                  2006
['Benjamin Blümchen']                                                  1503
['Bibi Blocksberg']                                                    1472
['Lata Mangeshkar']                                                    1373
                                                                       ... 
['Jess Penner']                                                           1
['Jess Ingerslev', 'Tritonus Koret']                                      1
['Jess Ingerslev', 'Bendt Reiner', 'Niels Weyde', 'Tritonus Koret']       1
['Jess Glynne', 'Jonas Blue']                                             1
['최진희']                                                                   1
Name: count, Length: 114030, dtype: int64

In [11]:
df_tracks["artists"].value_counts()

artists
['Die drei ???']                   3856
['TKKG Retro-Archiv']              2006
['Benjamin Blümchen']              1503
['Bibi Blocksberg']                1472
['Lata Mangeshkar']                1373
                                   ... 
['IU', 'Jang Yi-jeong']               1
['黃宗澤']                               1
['Vincy Chan', '海鳴威']                 1
['Dough-Boy']                         1
['Gentle Bones', 'Clara Benin']       1
Name: count, Length: 114030, dtype: int64

Artists Distribution Insights

Diverse Range: The artists column contains 114,030 unique entries, indicating a wide variety of artists.

Most Frequent Artist: The artist Die drei ??? has 3,856 tracks, suggesting significant popularity.

Collaborations: Some entries feature multiple artists, reflecting common collaborations in the music industry.

Analysis Potential: This distribution allows for further exploration of artist popularity, audio features, and genre diversity.

In [12]:
df_features.shape


(232725, 18)

In [13]:
df_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232724 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

In [14]:
df_features.isnull().sum()[df_features.isnull().sum()>0]

track_name    1
dtype: int64

In [15]:
df_features.duplicated().sum()[df_features.duplicated().sum()>0]

array([], dtype=int64)

In [16]:
df_features.value_counts("genre")

genre
Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9096
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: count, dtype: int64

In [17]:
# Rows null checked

df_tracks["name"].isnull().sum()

71

In [18]:
null_rows = df_tracks[df_tracks['name'].isnull()][['name', 'artists', 'popularity']]  # Adjust columns as needed
print(null_rows)

       name artists  popularity
226336  NaN    ['']          28
510975  NaN    ['']           0
510976  NaN    ['']           0
510978  NaN    ['']           0
510979  NaN    ['']           0
...     ...     ...         ...
517206  NaN    ['']           2
517215  NaN    ['']           2
520127  NaN    ['']           6
525238  NaN    ['']          26
525265  NaN    ['']          24

[71 rows x 3 columns]


In [19]:
df_tracks[df_tracks.isnull().any(axis = 1)]

id name  popularity  duration_ms  explicit  \
226336  4iH7negBYMfj2z0wDNmgdx  NaN          28       264973         0   
510975  04d5kbLvSAIBt3pGcljdhC  NaN           0       184293         0   
510976  05tRkgyxVdwMePGqOXMDYU  NaN           0       191587         0   
510978  0YAMRgAQH6tkTh4sWNXr8L  NaN           0       191573         0   
510979  1K6MQQxmFpPb66ZnaiIpHX  NaN           0       167602         0   
...                        ...  ...         ...          ...       ...   
517206  6OH9mz9aFbGlbf74cBwYWD  NaN           2       209760         0   
517215  15RqFDA86slfzujSQMEX4i  NaN           2       257280         0   
520127  0hKA9A2JPtFdg0fiMhyjQD  NaN           6       194081         0   
525238  1kR4gIb7nGxHPI3D2ifs59  NaN          26       289440         0   
525265  4xyBq8t4nkUKH8s28j6ZoD  NaN          24       254000         0   

       artists                  id_artists release_date  danceability  energy  \
226336    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1994-01-01         0.512   0.578   
510975    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1922-04-01         0.426   0.285   
510976    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1922-04-01         0.344   0.186   
510978    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1922-04-01         0.316   0.257   
510979    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1922-04-01         0.558   0.283   
...        ...                         ...          ...           ...     ...   
517206    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1962-02-01         0.506   0.598   
517215    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1962-02-01         0.612   0.615   
520127    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1974-12-31         0.471   0.369   
525238    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1998-01-05         0.501   0.583   
525265    ['']  ['0LyfQWJT6nXafLPZqxe9Of']   1998-01-05         0.390   0.452   

        key  loudness  mode  speechiness  acousticness  instrumentalness  \
226336    0   -12.280     0       0.0299        0.0433          0.000064   
510975   11   -11.970     1       0.0466        0.9950          0.264000   
510976    0   -13.495     1       0.0745        0.9950          0.000000   
510978    3   -13.611     0       0.0549        0.9950          0.769000   
510979    1   -12.847     1       0.0646        0.9960          0.000000   
...     ...       ...   ...          ...           ...               ...   
517206    7    -4.672     0       0.0635        0.7710          0.000001   
517215    5    -5.609     1       0.0551        0.8540          0.000023   
520127    4   -12.927     0       0.1460        0.9680          0.001100   
525238    7    -9.460     0       0.0605        0.6900          0.003960   
525265    7    -9.900     1       0.0361        0.7370          0.000000   

        liveness  valence    tempo  time_signature  
226336    0.5160    0.692  156.465               1  
510975    0.2930    0.583  135.661               4  
510976    0.1150    0.290   79.591               1  
510978    0.5190    0.529   68.682               3  
510979    0.4530    0.608   70.379               4  
...          ...      ...      ...             ...  
517206    0.6910    0.800   91.172               4  
517215    0.0541    0.809   90.536               4  
520127    0.1410    0.766   94.063               4  
525238    0.0747    0.734  138.391               4  
525265    0.1110    0.662   94.630               4  

[71 rows x 20 columns]

Given that there are 586,672 entries in the dataset, I have decided to drop the 71 null values in the name column. These null values represent tracks with empty artists and low popularity scores, contributing to their lack of meaningful data. Since they account for a very small proportion of the overall dataset, their absence will not significantly affect the analysis.

In [20]:
# Drop the missing Values
df_tracks.dropna(subset = ["name"],inplace = True)

In [21]:
# look the null values to other df and drop
df_features.isnull().sum()
df_features.dropna(subset = ["track_name"], inplace = True)

In [22]:
# Convert Data type
df_tracks.info()

<class 'pandas.core.frame.DataFrame'>
Index: 586601 entries, 0 to 586671
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                586601 non-null  object 
 1   name              586601 non-null  object 
 2   popularity        586601 non-null  int64  
 3   duration_ms       586601 non-null  int64  
 4   explicit          586601 non-null  int64  
 5   artists           586601 non-null  object 
 6   id_artists        586601 non-null  object 
 7   release_date      586601 non-null  object 
 8   danceability      586601 non-null  float64
 9   energy            586601 non-null  float64
 10  key               586601 non-null  int64  
 11  loudness          586601 non-null  float64
 12  mode              586601 non-null  int64  
 13  speechiness       586601 non-null  float64
 14  acousticness      586601 non-null  float64
 15  instrumentalness  586601 non-null  float64
 16  liveness          586601 

In [23]:
df_tracks.head()

id                                 name  popularity  \
0  35iwgR4jXetI318WEWsa1Q                                Carve           6   
1  021ht4sdgPcrDgSk7JTbKY  Capítulo 2.16 - Banquero Anarquista           0   
2  07A5yehtSnoedViJAZkNnc   Vivo para Quererte - Remasterizado           0   
3  08FmqUhxtyLTn6pAh6bk45        El Prisionero - Remasterizado           0   
4  08y9GfoqCWfOGsKdwojr5e                  Lady of the Evening           0   

   duration_ms  explicit              artists                  id_artists  \
0       126903         0              ['Uli']  ['45tIt06XoI0Iio4LBEVpls']   
1        98200         0  ['Fernando Pessoa']  ['14jtPCOoNZwquk5wd9DxrY']   
2       181640         0  ['Ignacio Corsini']  ['5LiOoJbxVSAMkBS2fUm3X2']   
3       176907         0  ['Ignacio Corsini']  ['5LiOoJbxVSAMkBS2fUm3X2']   
4       163080         0      ['Dick Haymes']  ['3BiJGZsyX9sJchTqcSA7Su']   

  release_date  danceability  energy  key  loudness  mode  speechiness  \
0   1922-02-22         0.645  0.4450    0   -13.338     1       0.4510   
1   1922-06-01         0.695  0.2630    0   -22.136     1       0.9570   
2   1922-03-21         0.434  0.1770    1   -21.180     1       0.0512   
3   1922-03-21         0.321  0.0946    7   -27.961     1       0.0504   
4         1922         0.402  0.1580    3   -16.900     0       0.0390   

   acousticness  instrumentalness  liveness  valence    tempo  time_signature  
0         0.674            0.7440     0.151    0.127  104.851               3  
1         0.797            0.0000     0.148    0.655  102.009               1  
2         0.994            0.0218     0.212    0.457  130.418               5  
3         0.995            0.9180     0.104    0.397  169.980               3  
4         0.989            0.1300     0.311    0.196  103.220               4

In [24]:
df_tracks["release_date"] = pd.to_datetime(df_tracks["release_date"].apply(lambda date: f"{date}-01-01" if len(str(date))==4 else date),errors = "coerce")

In [25]:
df_tracks.dtypes

id                          object
name                        object
popularity                   int64
duration_ms                  int64
explicit                     int64
artists                     object
id_artists                  object
release_date        datetime64[ns]
danceability               float64
energy                     float64
key                          int64
loudness                   float64
mode                         int64
speechiness                float64
acousticness               float64
instrumentalness           float64
liveness                   float64
valence                    float64
tempo                      float64
time_signature               int64
dtype: object

In [26]:
df_features.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [27]:
# Convert the 'mode' column to binary
df_features["mode_binary"] = df_features["mode"].map({"Minor":0, "Major":1})


In [28]:
# Define the mapping for the key column
key_mapping = {
    'C': 0,
    'C#': 1,
    'D': 2,
    'D#': 3,
    'E': 4,
    'F': 5,
    'F#': 6,
    'G': 7,
    'G#': 8,
    'A': 9,
    'A#': 10,
    'B': 11
}

df_features["key_numeric"] = df_features["key"].map(key_mapping)